resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/KubernetesExternalShuffleService.scala

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.kubernetes
+
+import java.nio.ByteBuffer
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.client.{KubernetesClientException, Watch, Watcher}
+import io.fabric8.kubernetes.client.Watcher.Action
+
+import org.apache.spark.{SecurityManager, SparkConf}
+import org.apache.spark.deploy.ExternalShuffleService
+import org.apache.spark.deploy.kubernetes.constants._
+import org.apache.spark.internal.Logging
+import org.apache.spark.network.client.{RpcResponseCallback, TransportClient}
+import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
+import org.apache.spark.network.shuffle.protocol.BlockTransferMessage
+import org.apache.spark.network.shuffle.protocol.kubernetes.RegisterKubernetesApp
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.scheduler.cluster.kubernetes.DriverPodKubernetesClientProvider
+
+
+/**
+ * An RPC endpoint that receives registration requests from Spark drivers running on Kubernetes.
+ * It detects driver termination and calls the cleanup callback to [[ExternalShuffleService]].
+ */
+private[spark] class KubernetesShuffleBlockHandler (
+  sparkConf: SparkConf,
+  transportConf: TransportConf)
+  extends ExternalShuffleBlockHandler(transportConf, null) with Logging {
+
+  private var connectedApps: Set[String] = Set()
+  private val CONNECTED_APPS_LOCK = new Object
+  private val Watch: Option[Watch] = ShuffleWatcher()
```

Comment:
I think that would need a change to make `ExternalShuffleService.blockHandler` a protected field, because we want to call `close()` on the [`blockHandler`](https://github.com/apache-spark-on-k8s/spark/blob/23e047f4609bf39f50ea4c65f704cac15408a821/core/src/main/scala/org/apache/spark/deploy/ExternalShuffleService.scala#L54) which is currently encapsulated within `ExternalShuffleClient`. The mesos external shuffle service creates a background thread during construction as well to avoid this I think.